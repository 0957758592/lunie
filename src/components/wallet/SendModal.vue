<template>
  <ActionModal
    id="send-modal"
    ref="actionModal"
    :submit-fn="submitForm"
    :simulate-fn="simulateForm"
    :validate="validateForm"
    :amount="amount"
    title="Send"
    submission-error-prefix="Sending tokens failed"
    @close="clear"
  >
    <TmFormGroup
      :error="$v.denom.$dirty && $v.denom.$invalid"
      class="action-modal-form-group"
      field-id="send-denomination"
      field-label="Denomination"
    >
      <TmField
        id="send-denomination"
        :value="num.viewDenom($v.denom.$model)"
        type="text"
        readonly
      />
      <TmFormMsg
        v-if="$v.denom.$error && !$v.denom.required"
        name="Denomination"
        type="required"
      />
    </TmFormGroup>

    <TmFormGroup
      :error="$v.address.$error && $v.address.$invalid"
      class="action-modal-form-group"
      field-id="send-address"
      field-label="Send To"
    >
      <TmField
        id="send-address"
        v-model.number="$v.address.$model"
        v-focus
        type="text"
        placeholder="Address"
      />
      <TmFormMsg
        v-if="$v.address.$error && !$v.address.required"
        name="Address"
        type="required"
      />
      <TmFormMsg
        v-else-if="$v.address.$error && !$v.address.bech32Validate"
        name="Address"
        type="bech32"
      />
    </TmFormGroup>
    <TmFormGroup
      :error="$v.amount.$error && $v.amount.$invalid"
      class="action-modal-form-group"
      field-id="amount"
      field-label="Amount"
    >
      <TmField
        id="amount"
        v-model="amount"
        class="tm-field"
        placeholder="Amount"
        type="number"
      />
      <TmFormMsg
        v-if="balance === 0"
        :msg="`doesn't have any ${num.viewDenom(denom)}s`"
        name="Wallet"
        type="custom"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && (!$v.amount.required || amount === 0)"
        name="Amount"
        type="required"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.decimal"
        name="Amount"
        type="numeric"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.between"
        :max="$v.amount.$params.between.max"
        :min="$v.amount.$params.between.min"
        name="Amount"
        type="between"
      />
    </TmFormGroup>
    <TmFormGroup
      :error="$v.memo.$error && $v.memo.$invalid"
      class="action-modal-group"
      field-id="memo"
      field-label="Memo"
    >
      <span class="input-suffix">
        {{ (memo ? " - " : "") + "(Sent via Lunie)" }}
      </span>
      <tm-field
        id="memo"
        v-model="memo"
        type="text"
        placeholder="Add a description..."
      />
      <tm-form-msg
        v-if="$v.memo.$error && !$v.memo.maxLength"
        name="Memo"
        type="maxLength"
        :max="max_memo_characters"
      />
    </TmFormGroup>
  </ActionModal>
</template>

<script>
import b32 from "scripts/b32"
import { required, between, decimal, maxLength } from "vuelidate/lib/validators"
import num, { uatoms, atoms, SMALLEST } from "../../scripts/num.js"
import { mapActions, mapGetters } from "vuex"
import TmFormGroup from "common/TmFormGroup"
import TmField from "common/TmField"
import TmFormMsg from "common/TmFormMsg"
import ActionModal from "common/ActionModal"

export default {
  name: `send-modal`,
  components: {
    TmField,
    TmFormGroup,
    TmFormMsg,
    ActionModal
  },
  data: () => ({
    address: ``,
    amount: null,
    denom: ``,
    num,
    memo: null,
    max_memo_characters: 256 - ` + (Sent via Lunie)`.length
  }),
  computed: {
    ...mapGetters([`wallet`]),
    balance() {
      const denom = this.wallet.balances.find(b => b.denom === this.denom)
      return (denom && denom.amount) || 0
    }
  },
  mounted() {
    if (this.denom) {
      this.denom = this.denom
    }
  },
  methods: {
    ...mapActions([`sendTx`]),
    open(denom) {
      this.denom = denom
      this.$refs.actionModal.open()
    },
    validateForm() {
      this.$v.$touch()

      return !this.$v.$invalid
    },
    clear() {
      this.$v.$reset()

      this.address = ``
      this.amount = 0
    },
    async simulateForm() {
      const amount = +this.amount
      const address = this.address
      const denom = this.denom
      const type = `send`

      return await this.$store.dispatch(`simulateTx`, {
        type,
        to: address,
        amount: [{ denom, amount: String(uatoms(amount)) }],
        memo: this.memo
      })
    },
    async submitForm(gasEstimate, gasPrice, password, submitType) {
      const amount = +this.amount
      const address = this.address
      const denom = this.denom
      const type = `send`

      await this.sendTx({
        type,
        gas: String(gasEstimate),
        gas_prices: [
          {
            amount: String(uatoms(gasPrice)),
            denom: this.denom // TODO: should always match staking denom
          }
        ],
        submitType,
        password,
        to: address,
        amount: [{ denom, amount: String(uatoms(amount)) }],
        memo: this.memo
      })

      this.$store.commit(`notify`, {
        title: `Successful Send`,
        body: `Successfully sent ${amount} ${num.viewDenom(
          denom
        )}s to ${address}`
      })
    },
    bech32Validate(param) {
      try {
        b32.decode(param)
        return true
      } catch (error) {
        return false
      }
    }
  },
  validations() {
    return {
      address: {
        required,
        bech32Validate: this.bech32Validate
      },
      amount: {
        required: x => !!x && x !== `0`,
        decimal,
        between: between(SMALLEST, atoms(this.balance))
      },
      denom: { required },
      memo: {
        maxLength: maxLength(
          this.max_memo_characters - ` + (Sent via Lunie)`.length
        )
      }
    }
  }
}
</script>
