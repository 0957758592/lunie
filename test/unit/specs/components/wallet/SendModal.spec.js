import { shallowMount, createLocalVue } from "@vue/test-utils"
import Vuelidate from "vuelidate"
import SendModal from "renderer/components/wallet/SendModal"

describe(`SendModal`, () => {
  const localVue = createLocalVue()
  localVue.use(Vuelidate)
  localVue.directive(`focus`, () => { })

  let wrapper, $store

  const balances = [
    {
      denom: `STAKE`,
      amount: 10000000000
    },
    {
      denom: `fermion`,
      amount: 2300
    }
  ]
  const getters = {
    wallet: {
      loading: false,
      denoms: [`fermion`, `gregcoin`, `mycoin`, `STAKE`],
      balances
    },
    connected: true,
    session: { signedIn: true }
  }

  beforeEach(async () => {
    $store = {
      commit: jest.fn(),
      dispatch: jest.fn(),
      getters
    }

    wrapper = shallowMount(SendModal, {
      localVue,
      propsData: {
        denom: `STAKE`
      },
      mocks: {
        $store
      },
      sync: false
    })

    wrapper.vm.$refs.actionModal = {
      submit: cb => cb(),
      open: jest.fn()
    }
  })

  it(`should display send modal form`, async () => {
    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  it(`clears on close`, () => {
    const $v = { $reset: jest.fn() }
    const address = `cosmos1address`
    const amount = 10
    SendModal.methods.clear.call($v, address, amount)
    expect($v.$reset).toHaveBeenCalled()
    expect(address).toBe(``)
    expect(amount).toBe(0)
  })

  describe(`validation`, () => {
    it(`should show address required error`, async () => {
      wrapper.setData({
        denom: `STAKE`,
        address: ``,
        amount: 2
      })
      wrapper.vm.validateForm()
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.$v.$error).toBe(true)
      expect(wrapper.vm.$el).toMatchSnapshot()
    })
    it(`should show bech32 error when address length is too short`, async () => {
      wrapper.setData({
        denom: `STAKE`,
        address: `asdf`,
        amount: 2
      })
      wrapper.vm.validateForm()
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.$v.$error).toBe(true)
      expect(wrapper.vm.$el).toMatchSnapshot()
    })

    it(`should show bech32 error when address length is too long`, async () => {
      wrapper.setData({
        denom: `STAKE`,
        address: `asdfasdfasdfasdfasdfasdfasdfasdfasdfasdfasdf`,
        amount: 2
      })
      wrapper.vm.validateForm()
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.$v.$error).toBe(true)
      expect(wrapper.vm.$el).toMatchSnapshot()
    })
    it(`should show bech32 error when alphanumeric is wrong`, async () => {
      wrapper.setData({
        address: ``
      })
      expect(wrapper.vm.validateForm()).toBe(false)
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.$el).toMatchSnapshot()
    })
  })

  describe(`submitForm`, () => {
    it(`submits a transfer transaction`, async () => {
      const $store = { commit: jest.fn() }
      const sendTx = jest.fn()

      await SendModal.methods.submitForm.call(
        {
          amount: 10,
          address: `cosmos1address`,
          denom: `uatom`,
          $store,
          sendTx
        },
        `ledger`, ``
      )

      expect(sendTx).toHaveBeenCalledWith(`submitProposal`,
        {
          type: `send`,
          to: `cosmos1address`,
          amount: [{ amount: `15000000`, denom: `uatom` }],
          submitType: `ledger`,
          password: ``
        }
      )

      expect($store.commit).toHaveBeenCalledWith(`notify`,
        {
          body: `You have successfully submitted a new text proposal`,
          title: `Successful proposal submission!`
        }
      )
    })
  })
  it(`validates bech32 addresses`, () => {
    expect(
      wrapper.vm.bech32Validate(`cosmos1x7wzdumfj8pncd99mqc0mkqfrrps3l3pjz8tk6`)
    ).toBe(true)
    expect(
      wrapper.vm.bech32Validate(`cosmos15ky9du8a2wlstz6fpx3p4mqpjyrm5ctpesxxn9`)
    ).toBe(false)
  })
})
