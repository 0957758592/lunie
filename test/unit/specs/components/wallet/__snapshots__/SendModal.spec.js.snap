// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendModal disables sending if not connected 1`] = `
<div
  class="action-modal"
  name="slide-fade"
>
  <div
    class="action-modal-header"
  >
    <img
      class="icon action-modal-atom"
      src="~assets/images/cosmos-logo.png"
    />
    <span
      class="action-modal-title"
    >
      Send
    </span>
     
    <div
      class="action-modal-icon action-modal-close"
      id="closeBtn"
    >
      <i
        class="material-icons"
      >
        close
      </i>
    </div>
  </div>
   
  <div
    class="action-modal-form"
  >
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-denomination"
      >
        Denomination
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <div
          class="tm-select"
          id="send-denomination"
        >
          <select
            class="tm-field tm-field-select"
          >
            <option
              disabled="disabled"
              hidden="hidden"
              selected="selected"
              value=""
            >
              Select token...
            </option>
             
            <option
              value="mycoin"
            >
              MYCOIN
            </option>
            <option
              value="fermion"
            >
              FERMION
            </option>
          </select>
           
          <div
            class="tm-field-select-addon"
          >
            <i
              class="material-icons"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-address"
      >
        Send To
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <div
          class="tm-field-group"
        >
          <input
            class="tm-field"
            id="send-address"
            placeholder="Address"
            type="text"
          />
        </div>
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-amount"
      >
        Amount
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <div
          class="tm-field-group"
        >
          <input
            class="tm-field"
            id="send-amount"
            max="2300"
            min="1"
            placeholder="Amount"
            type="number"
          />
        </div>
         
        <!---->
         
        <!---->
      </div>
    </div>
     
    <!---->
     
    <!---->
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="password"
      >
        Password
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="password"
          placeholder="Password"
          type="password"
        />
         
        <!---->
      </div>
    </div>
     
  </div>
   
  <div
    class="action-modal-footer"
  >
    <div>
      <button
        class="tm-btn"
        disabled="disabled"
      >
        <span
          class="tm-btn__container tm-btn--primary"
        >
          <!---->
          <span
            class="tm-btn__value"
          >
            Connecting...
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`SendModal has the expected html structure 1`] = `
<div
  class="action-modal"
  name="slide-fade"
>
  <div
    class="action-modal-header"
  >
    <img
      class="icon action-modal-atom"
      src="~assets/images/cosmos-logo.png"
    />
    <span
      class="action-modal-title"
    >
      Send
    </span>
     
    <div
      class="action-modal-icon action-modal-close"
      id="closeBtn"
    >
      <i
        class="material-icons"
      >
        close
      </i>
    </div>
  </div>
   
  <div
    class="action-modal-form"
  >
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-denomination"
      >
        Denomination
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <div
          class="tm-select"
          id="send-denomination"
        >
          <select
            class="tm-field tm-field-select"
          >
            <option
              disabled="disabled"
              hidden="hidden"
              selected="selected"
              value=""
            >
              Select token...
            </option>
             
            <option
              value="mycoin"
            >
              MYCOIN
            </option>
            <option
              value="fermion"
            >
              FERMION
            </option>
          </select>
           
          <div
            class="tm-field-select-addon"
          >
            <i
              class="material-icons"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group tm-form-group--error"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-address"
      >
        Send To
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <div
          class="tm-field-group"
        >
          <input
            class="tm-field"
            id="send-address"
            placeholder="Address"
            type="text"
          />
        </div>
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-amount"
      >
        Amount
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <div
          class="tm-field-group"
        >
          <input
            class="tm-field"
            id="send-amount"
            max="2300"
            min="1"
            placeholder="Amount"
            type="number"
          />
        </div>
         
        <!---->
         
        <!---->
      </div>
    </div>
     
    <!---->
     
    <!---->
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="password"
      >
        Password
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="password"
          placeholder="Password"
          type="password"
        />
         
        <div
          class="tm-form-msg sm tm-form-msg--error"
        >
          Password is required
        </div>
      </div>
    </div>
     
  </div>
   
  <div
    class="action-modal-footer"
  >
    <div>
      <button
        class="tm-btn"
        id="send-btn"
      >
        <span
          class="tm-btn__container tm-btn--primary"
        >
          <!---->
          <span
            class="tm-btn__value"
          >
            Send Tokens
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`SendModal should show bech32 error when address length is too long 1`] = `
<div
  class="action-modal"
  name="slide-fade"
>
  <div
    class="action-modal-header"
  >
    <img
      class="icon action-modal-atom"
      src="~assets/images/cosmos-logo.png"
    />
    <span
      class="action-modal-title"
    >
      Send
    </span>
     
    <div
      class="action-modal-icon action-modal-close"
      id="closeBtn"
    >
      <i
        class="material-icons"
      >
        close
      </i>
    </div>
  </div>
   
  <div
    class="action-modal-form"
  >
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-denomination"
      >
        Denomination
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <div
          class="tm-select"
          id="send-denomination"
        >
          <select
            class="tm-field tm-field-select"
          >
            <option
              disabled="disabled"
              hidden="hidden"
              selected="selected"
              value=""
            >
              Select token...
            </option>
             
            <option
              value="mycoin"
            >
              MYCOIN
            </option>
            <option
              value="fermion"
            >
              FERMION
            </option>
          </select>
           
          <div
            class="tm-field-select-addon"
          >
            <i
              class="material-icons"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group tm-form-group--error"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-address"
      >
        Send To
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <div
          class="tm-field-group"
        >
          <input
            class="tm-field"
            id="send-address"
            placeholder="Address"
            type="text"
          />
        </div>
         
        <div
          body="No separator character for asdfasdfasdfasdfasdfasdfasdfasdfasdfasdfasdf"
          class="tm-form-msg sm tm-form-msg--error"
        >
          Address is invalid bech32
        </div>
      </div>
    </div>
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-amount"
      >
        Amount
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <div
          class="tm-field-group"
        >
          <input
            class="tm-field"
            id="send-amount"
            max="2300"
            min="1"
            placeholder="Amount"
            type="number"
          />
        </div>
         
        <!---->
         
        <!---->
      </div>
    </div>
     
    <!---->
     
    <!---->
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="password"
      >
        Password
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="password"
          placeholder="Password"
          type="password"
        />
         
        <!---->
      </div>
    </div>
     
  </div>
   
  <div
    class="action-modal-footer"
  >
    <div>
      <button
        class="tm-btn"
        disabled="disabled"
      >
        <span
          class="tm-btn__container tm-btn--primary"
        >
          <!---->
          <span
            class="tm-btn__value"
          >
            Sending...
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`SendModal should show bech32 error when address length is too short 1`] = `
<div
  class="action-modal"
  name="slide-fade"
>
  <div
    class="action-modal-header"
  >
    <img
      class="icon action-modal-atom"
      src="~assets/images/cosmos-logo.png"
    />
    <span
      class="action-modal-title"
    >
      Send
    </span>
     
    <div
      class="action-modal-icon action-modal-close"
      id="closeBtn"
    >
      <i
        class="material-icons"
      >
        close
      </i>
    </div>
  </div>
   
  <div
    class="action-modal-form"
  >
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-denomination"
      >
        Denomination
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <div
          class="tm-select"
          id="send-denomination"
        >
          <select
            class="tm-field tm-field-select"
          >
            <option
              disabled="disabled"
              hidden="hidden"
              selected="selected"
              value=""
            >
              Select token...
            </option>
             
            <option
              value="mycoin"
            >
              MYCOIN
            </option>
            <option
              value="fermion"
            >
              FERMION
            </option>
          </select>
           
          <div
            class="tm-field-select-addon"
          >
            <i
              class="material-icons"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group tm-form-group--error"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-address"
      >
        Send To
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <div
          class="tm-field-group"
        >
          <input
            class="tm-field"
            id="send-address"
            placeholder="Address"
            type="text"
          />
        </div>
         
        <div
          body="asdf too short"
          class="tm-form-msg sm tm-form-msg--error"
        >
          Address is invalid bech32
        </div>
      </div>
    </div>
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-amount"
      >
        Amount
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <div
          class="tm-field-group"
        >
          <input
            class="tm-field"
            id="send-amount"
            max="2300"
            min="1"
            placeholder="Amount"
            type="number"
          />
        </div>
         
        <!---->
         
        <!---->
      </div>
    </div>
     
    <!---->
     
    <!---->
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="password"
      >
        Password
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="password"
          placeholder="Password"
          type="password"
        />
         
        <!---->
      </div>
    </div>
     
  </div>
   
  <div
    class="action-modal-footer"
  >
    <div>
      <button
        class="tm-btn"
        disabled="disabled"
      >
        <span
          class="tm-btn__container tm-btn--primary"
        >
          <!---->
          <span
            class="tm-btn__value"
          >
            Sending...
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`SendModal should show bech32 error when alphanumeric is wrong 1`] = `
<div
  class="action-modal"
  name="slide-fade"
>
  <div
    class="action-modal-header"
  >
    <img
      class="icon action-modal-atom"
      src="~assets/images/cosmos-logo.png"
    />
    <span
      class="action-modal-title"
    >
      Send
    </span>
     
    <div
      class="action-modal-icon action-modal-close"
      id="closeBtn"
    >
      <i
        class="material-icons"
      >
        close
      </i>
    </div>
  </div>
   
  <div
    class="action-modal-form"
  >
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-denomination"
      >
        Denomination
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <div
          class="tm-select"
          id="send-denomination"
        >
          <select
            class="tm-field tm-field-select"
          >
            <option
              disabled="disabled"
              hidden="hidden"
              selected="selected"
              value=""
            >
              Select token...
            </option>
             
            <option
              value="mycoin"
            >
              MYCOIN
            </option>
            <option
              value="fermion"
            >
              FERMION
            </option>
          </select>
           
          <div
            class="tm-field-select-addon"
          >
            <i
              class="material-icons"
            >
              arrow_drop_down
            </i>
          </div>
        </div>
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group tm-form-group--error"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-address"
      >
        Send To
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <div
          class="tm-field-group"
        >
          <input
            class="tm-field"
            id="send-address"
            placeholder="Address"
            type="text"
          />
        </div>
         
        <div
          body="No separator character for !@#$!@#$!@#$!@#$!@#$!@#$!@#$!@#$!@#$!@#$"
          class="tm-form-msg sm tm-form-msg--error"
        >
          Address is invalid bech32
        </div>
      </div>
    </div>
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-amount"
      >
        Amount
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <div
          class="tm-field-group"
        >
          <input
            class="tm-field"
            id="send-amount"
            max="2300"
            min="1"
            placeholder="Amount"
            type="number"
          />
        </div>
         
        <!---->
         
        <!---->
      </div>
    </div>
     
    <!---->
     
    <!---->
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="password"
      >
        Password
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="password"
          placeholder="Password"
          type="password"
        />
         
        <!---->
      </div>
    </div>
     
  </div>
   
  <div
    class="action-modal-footer"
  >
    <div>
      <button
        class="tm-btn"
        disabled="disabled"
      >
        <span
          class="tm-btn__container tm-btn--primary"
        >
          <!---->
          <span
            class="tm-btn__value"
          >
            Sending...
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
